plugins/MumblePluginProcess.hpp

```diff
@@ -0,0 +1,313 @@
+		try {
+			std::vector< T > var(elements);
+			peek(address, &var[0], sizeof(T) * elements);
+			return var;
+		} catch (std::bad_alloc &) {
+			return std::vector< T >();
+		}
+	}
+
+	/// This function is very useful when there is an array of structures and we only need to store a part of each.
+	/// The Source Engine plugin needs it because there is a structure that happens to be larger in the Linux version of
+	/// Left 4 Dead 2. Since the extra members are unknown, we simply discard the extra bytes when reading the array of
+	/// structures.
+	template< typename T >
+	std::vector< T > peekVector(const procptr_t address, const size_t elements, const size_t realStructSize) {
+		if (realStructSize == sizeof(T)) {
+			// If the structure's size is correct, there is no need for this special function.
+			return peekVector< T >(address, elements);
+		} else if (realStructSize < sizeof(T)) {
+			return std::vector< T >();
+		}
+
+		// Read the full structures into a vector.
+		const auto fullStructs = peekVector< uint8_t >(address, realStructSize * elements);
+
+		// Allocate vector for the known structures.
+		std::vector< T > ret(elements);
+
+		// Copy elements from the vector containing the full structures, discarding extra bytes.
+		auto seek = &fullStructs[0];
+
+		for (size_t i = 0; i < elements; ++i, seek += realStructSize) {
+			memcpy(&ret[i], seek, sizeof(T));
+		}
+
+		return ret;
+	}
+
+	/// Reads the specified amount of data at the specified address and returns it as std::string.
+	/// An empty std::string is returned in case of error.
+	///
+	/// If \p length is 0, the function reads one byte at a time and stops when either '\0' is found or 3 seconds have
+	/// passed. The successfully read data is returned, also in case of error.
+	std::string peekString(const procptr_t address, const size_t length = 0) {
+		std::string string;
+
+		if (length > 0) {
+			string.resize(length);
+
+			if (!peek(address, &string[0], length)) {
+				return std::string();
+			}
+		} else {
+			auto now       = std::chrono::steady_clock::now();
+			const auto end = now + std::chrono::seconds(3);
+
+			for (procptr_t i = 0; now < end; ++i) {
+				char ch = 0;
+				if (!peek(address + i, &ch, sizeof(ch)) || ch == '\0') {
+					break;
+				}
+
+				string += ch;
+
+				// Update current time.
+				now = std::chrono::steady_clock::now();
+			}
+		}
+
+		return string;
+	}
+
+	procptr_t peekPtr(const procptr_t &addr) {
+		procptr_t v = 0;
+
+		if (!peek(addr, &v, pointerSize)) {
+			return 0;
+		}
+
+		return v;
+	}
+
+	procptr_t virtualFunction(const procptr_t classObject, const size_t index) {
+		const auto vTable = peekPtr(classObject);
+		if (!vTable) {
+			return 0;
+		}
+
+		return peekPtr(vTable + (index * pointerSize));
+	}
+
+	inline procptr_t module() { return module(name); }
+#ifdef OS_WINDOWS
+	virtual procptr_t module(const std::string &module) = 0;
+#else
+	procptr_t module(const std::string &module) {
```

Comment:
Again: This should be single pure virtual function that is implemented by the respective usb-class

Reply:
The reason why I didn't do it is due to `module()`'s implementation being the same for both native and Wine processes.